core/server/pagination.go

```diff
@@ -0,0 +1,79 @@
+package server
+
+import (
+	"fmt"
+
+	v1 "k8s.io/api/core/v1"
+)
+
+type PageTokenInfo struct {
+	NamespaceIndex int    `json:"namespace_index"`
+	Namespace      string `json:"namespace"`
+	K8sPageToken   string `json:"k8s_page_token"`
+}
+
+type getRawPageData func(namespace string, limit int32, pageToken string) (string, int32, error)
+
+func GetNextPage(namespaceList []v1.Namespace, pageSize int32, pageToken string, getRawPage getRawPageData) (string, error) {
+	newPageToken := ""
+	currentNamespaceIndex := 0
+	currentK8sPageToken := ""
+	itemsLeft := pageSize
+
+	if pageToken != "" {
+		var pageTokenInfo PageTokenInfo
+
+		err := decodeFromBase64(&pageTokenInfo, pageToken)
+		if err != nil {
+			return "", fmt.Errorf("error decoding next token %w", err)
+		}
+
+		currentNamespaceIndex = pageTokenInfo.NamespaceIndex
+		currentK8sPageToken = pageTokenInfo.K8sPageToken
+	}
+
+	for nsIndex, ns := range namespaceList[currentNamespaceIndex:] {
+		nextPageToken, pageLen, err := getRawPage(ns.Name, itemsLeft, currentK8sPageToken)
+		if err != nil {
+			continue
```

Comment:
yes

Reply:
actually we don't, this error is being logged already inside `getRawPage`.

Reply:
This logic is a little confusing to me: "If there is an error skip to the next item". We don't need to do anything with `itemsLeft` or `currentK8sPageToken`?

I'm worried this will lead to silent bugs. Maybe the logging should be moved to here?

Reply:
Actually, you are right. If an error occurs here I need to empty `currentK8sPageToken`. But this is only fine if we are ok ignoring the error (or just logging it ). At this point, I'm not too sure, as this might lead to inconsistent information or as you said silent bugs. I wonder if we should make the request fail if we find an error here instead 🤔. If so then we won't need to reset `currentK8sPageToken`.

Reply:
I think it would be better to address this issue as part of a different PR.